### PR TITLE
fix: sync enum trait rendering with enum shape as possible

### PR DESCRIPTION
**Pull request by i10416**
Previously, rendering result of [enum trait](https://smithy.io/2.0/spec/constraint-traits.html#enum-trait) was much different from that of [enum shape](https://smithy.io/2.0/spec/simple-types.html#enum) in that
1. smithy4s dropped enum variant hints
2. sealed abstract classes generated from enum trait held `smithy.api#enum` hints
3. sealed abstract classes from enum trait didn't have `@param` properties generated.

For example, the following smithy definition generated the Scala code below.

```smithy
@enum([
{
    value: "HEAD",
    name: "HEAD",
    documentation: "head"
},
{
    value: "TAIL",
    name: "TAIL",
    documentation: "tail",
    deprecated: true
}
])

string Coin
```

```scala
// When Coin is generated from enum shape and enum variant has
// documentation, here are `@param` properties for each variant with doc.
sealed abstract class Coin ...

...

object Coin extends Enumeration[Suit] with ...
  ...
  val hints: Hints = Hints(
    smithy.api.Enum(...) // <- When Coin is generated from enum shape, smithy.api.Enum does not exist here.
  )

  /** head */
  case object HEAD extends Coin("HEAD", "HEAD", 0, Hints()) // <- Hints is not empty when Coin is enum shape.
  /** tail  */
  @deprecated
  case object TAIL extends Coin("TAIL", "TAIL", 1, Hints()) // too

```

After this commits, enum variants generated from enum trait also hold hints such as documentation or deprecation. In addition, sealed abstract classes generated from enum trait no longer have redundant `smithy.api#enum hints`.

See https://github.com/disneystreaming/smithy4s/pull/980/files#diff-1374f8e9776053385d67f4bc78d81b77bd7730a72ebd38801479005fba548fa0 for instance.

Note that enum trait and enum shape are not 100% compatible.
For example, enum trait support neither `deprecated$since` nor `deprecated$message`.

### CODE REVIEW

File: modules/codegen/src/smithy4s/codegen/internals/SmithyToIR.scala [hunk 1]
```diff
@@ -311,31 +311,19 @@ private[codegen] class SmithyToIR(model: Model, namespace: String) {
...
-              .map { case (value, index) =>
-                EnumValue(
-                  value.getValue(),
-                  index,
-                  EnumUtil.enumValueClassName(
-                    value.getName().asScala,
-                    value.getValue,
-                    index
-                  ),
-                  hints = Nil
```
Comment (i10416) [inline modules/codegen/src/smithy4s/codegen/internals/SmithyToIR.scala:328]: This dropped all variants' hints even if there were documentation or deprecation although https://github.com/disneystreaming/smithy4s/pull/980/files#diff-3d87f30f2b03db7c8e012d2c726b1e1a0f1227fe70d2e5c24166a9046c4428a5L337 preserved documentation and deprecation.

File: modules/codegen/test/src/smithy4s/codegen/internals/RendererSpec.scala [hunk 2]
```diff
@@ -85,6 +85,144 @@ final class RendererSpec extends munit.FunSuite {
...
+        )
+      ) match {
+        case None =>
+          fail(
+            "enum must be rendered as a sealed abstract class and its companion object"
+          )
+        case Some(code) => code
+      }
+    val classDoc = """/** this is an enum Suit
+                     |  * @param DIAMOND
```
Comment (i10416) [inline modules/codegen/test/src/smithy4s/codegen/internals/RendererSpec.scala:137]: Random thought: would it be better to generate `@param <variant>` even when there's no doc for variant?

Comment (i10416) [inline modules/codegen/test/src/smithy4s/codegen/internals/RendererSpec.scala:137]: Random thought: would it be better to generate `@param <variant>` even when there's no doc for variant?

File: modules/codegen/test/src/smithy4s/codegen/internals/RendererSpec.scala [hunk 3]
```diff
@@ -85,6 +85,144 @@ final class RendererSpec extends munit.FunSuite {
...
+                     |  *   this is a DIAMOND
+                     |  * @param HAERT
+                     |  *   this is a HAERT
+                     |  * @param HEART
+                     |  *   this is a HEART
+                     |  */""".stripMargin
+    val diamondWithDocRendered =
```
Comment (i10416) [inline modules/codegen/test/src/smithy4s/codegen/internals/RendererSpec.scala:144]: confirm variants holds hints from `@enum` trait

**Comment by i10416** [comment]
Hmm, I missed some corner cases.

File: modules/codegen/test/src/smithy4s/codegen/internals/RendererSpec.scala (hunk 2 above)
Comment (i10416) [reply modules/codegen/test/src/smithy4s/codegen/internals/RendererSpec.scala:137]: Hmm, as mentioned https://github.com/disneystreaming/smithy4s/issues/773, `@param` sounds strange for enum variants.

File: modules/codegen/src/smithy4s/codegen/internals/SmithyToIR.scala [hunk 4]
```diff
@@ -311,31 +311,55 @@ private[codegen] class SmithyToIR(model: Model, namespace: String) {
...
+                      shape
+                        .getAllTraits()
+                        .values()
+                        .asScala
+                        .filterNot(
+                          _.toShapeId() == ShapeId.from("smithy.api#enum")
+                        )
+                        .asJavaCollection
+                    )
+                    .asInstanceOf[EnumShape.Builder]
```
Comment (kubukoz) [inline modules/codegen/src/smithy4s/codegen/internals/SmithyToIR.scala:360]: **suggestion:** can you use an upcast instead? (`EnumShape.builder().(...): EnumShape.Builder`)

File: modules/example/src/smithy4s/example/LowHigh.scala [hunk 5]
```diff
@@ -20,12 +25,12 @@ sealed abstract class LowHigh(_value: String, _name: String, _intValue: Int, _hi
 object LowHigh extends Enumeration[LowHigh] with ShapeTag.Companion[LowHigh] {
   val id: ShapeId = ShapeId("smithy4s.example", "LowHigh")
 
-  val hints: Hints = Hints(
-    smithy.api.Enum(List(smithy.api.EnumDefinition(value = smithy.api.NonEmptyString("Low"), name = Some(smithy.api.EnumConstantBodyName("LOW")), documentation = None, tags = None, deprecated = None), smithy.api.EnumDefinition(value = smithy.api.NonEmptyString("High"), name = Some(smithy.api.EnumConstantBodyName("HIGH")), documentation = None, tags = None, deprecated = None))),
-  )
+  val hints: Hints = Hints.empty
```
Comment (kubukoz) [inline modules/example/src/smithy4s/example/LowHigh.scala:28]: **issue:** this doesn't look right, IMO we should keep the traits originally present in the model, regardless if they're duplicated in the Enumeration schema

File: modules/codegen/src/smithy4s/codegen/internals/SmithyToIR.scala (hunk 4 above)
Comment (i10416) [reply modules/codegen/src/smithy4s/codegen/internals/SmithyToIR.scala:360]: I think it gets compile error as `source` and `addTraits` come from `StringShape.Builder`, which is a super class of `EnumShape.Builder`.

File: modules/example/src/smithy4s/example/LowHigh.scala (hunk 5 above)
Comment (i10416) [reply modules/example/src/smithy4s/example/LowHigh.scala:28]: I found `fromStringShape` in software.amazon.smithy.model.shapes.EnumShape.java drops enum trait from `StringShape`. Do we still need to keep it? 


```java
    public static Optional<EnumShape> fromStringShape(StringShape shape, boolean synthesizeNames) {
        if (shape.isEnumShape()) {
            return Optional.of((EnumShape) shape);
        }

        if (!shape.hasTrait(EnumTrait.ID)) {
            return Optional.empty();
        }
        StringShape stringWithoutEnumTrait = shape.toBuilder().removeTrait(EnumTrait.ID).build();
        Builder enumBuilder = EnumShape.builder();
        stringWithoutEnumTrait.updateBuilder(enumBuilder);
        ...
```

Comment (i10416) [reply modules/example/src/smithy4s/example/LowHigh.scala:28]: https://github.com/awslabs/smithy/blob/main/smithy-model/src/main/java/software/amazon/smithy/model/shapes/EnumShape.java#L131

Comment (kubukoz) [reply modules/example/src/smithy4s/example/LowHigh.scala:28]: I guess there's no "right thing to do", it'll be a judgement call - do we try to do what Smithy does in `upgradeEnums` (which is an optional transformation that uses the code path you mentioned), or do we try to make our representation as close to the original as possible?

I think for support of Smithy 1.x, we usually did the latter.

**Comment by kubukoz** [comment]
Request: please avoid force pushes on PRs that have already been seen :) it's fine to keep "bad" commits on the branches, as we will squash them anyway.

**Comment by i10416** [comment]
So sorry for that 🙇

File: modules/example/src/smithy4s/example/LowHigh.scala (hunk 5 above)
Comment (i10416) [reply modules/example/src/smithy4s/example/LowHigh.scala:28]: Okey, I will keep enum trait information in generated hints.

File: modules/codegen/src/smithy4s/codegen/internals/SmithyToIR.scala [hunk 6]
```diff
@@ -311,31 +311,66 @@ private[codegen] class SmithyToIR(model: Model, namespace: String) {
...
+                throw new Exception(
+                  "smithy.api#enum shape is missing for enum trait"
                 )
+              )
+            val pseudoEnumShape =
+              EnumShape.fromStringShape(shape, true).asScala match {
+                case Some(shape) =>
+                  shape
+                    .toBuilder()
+                    .addTrait(enumTrait)
```
Comment (i10416) [inline modules/codegen/src/smithy4s/codegen/internals/SmithyToIR.scala]: restore `EnumTrait` dropped by `fromStringShape`

File: modules/codegen/src/smithy4s/codegen/internals/SmithyToIR.scala [hunk 7]
```diff
@@ -311,31 +311,66 @@ private[codegen] class SmithyToIR(model: Model, namespace: String) {
...
+                              .build()
+                        }
+                    }
+                    val builder = e.toBuilder().clearEnums()
+                    defs.foreach(builder.addEnum)
+                    builder.build()
+                  }
+                  EnumShape
+                    .builder()
+                    .addTrait(enumTrait)
```
Comment (i10416) [inline modules/codegen/src/smithy4s/codegen/internals/SmithyToIR.scala]: restore `EnumTrait` dropped by `fromStringShape`

File: modules/codegen/test/src/smithy4s/codegen/internals/RendererSpec.scala (hunk 2 above)
Comment (Baccata) [reply modules/codegen/test/src/smithy4s/codegen/internals/RendererSpec.scala:137]: Good point, maybe bullet points might be better in this case.

File: modules/example/src/smithy4s/example/LowHigh.scala (hunk 5 above)
Comment (Baccata) [reply modules/example/src/smithy4s/example/LowHigh.scala:28]: A little late to the party but I think it's fine to omit this trait. All semantic information is recoverable from the Schema, and this trait just bloats the generated code (and therefore bytecode) for no reason. 

If there was a reasonably good reason for advocating for its presence, I'd be inclined to keep it, but in the absence of such an argument, let's remove the generated `enum` hint. 

PS : sorry for the churn.

File: modules/codegen/src/smithy4s/codegen/internals/SmithyToIR.scala (hunk 6 above)
Comment (Baccata) [reply modules/codegen/src/smithy4s/codegen/internals/SmithyToIR.scala]: See my comment here : https://github.com/disneystreaming/smithy4s/pull/980#discussion_r1214054302

File: modules/codegen/src/smithy4s/codegen/internals/SmithyToIR.scala (hunk 7 above)
Comment (Baccata) [reply modules/codegen/src/smithy4s/codegen/internals/SmithyToIR.scala]: See my comment here : https://github.com/disneystreaming/smithy4s/pull/980#discussion_r1214054302

File: modules/codegen/test/src/smithy4s/codegen/internals/RendererSpec.scala [hunk 8]
```diff
@@ -85,6 +85,192 @@ final class RendererSpec extends munit.FunSuite {
...
+    assert(
+      definition.contains(diamondWithDocRendered),
+      "DIAMOND variant must hold hints from `@enum` trait"
+    )
+    assert(
+      haertWithDeprecationAndDocRendered.forall(definition.contains),
+      "variants generated from `@enum` trait must hold hints, but deprecation hint cannot support message and since properties."
+    )
+    assert(
+      !definition.contains("smithy.api.Enum"),
```
Comment (i10416) [inline modules/codegen/test/src/smithy4s/codegen/internals/RendererSpec.scala:166]: updated: generated shape doesn't have redundant metadata.

File: modules/codegen/test/src/smithy4s/codegen/internals/RendererSpec.scala [hunk 9]
```diff
@@ -85,6 +85,192 @@ final class RendererSpec extends munit.FunSuite {
...
+                     |  *   this is a DIAMOND
+                     |  * @param HAERT
+                     |  *   this is a HAERT
+                     |  * @param HEART
+                     |  *   this is a HEART
+                     |  */""".stripMargin
+    val diamondWithDocRendered =
+      """case object DIAMOND extends Suit("DIAMOND", "DIAMOND", 0, Hints(smithy.api.Documentation("this is a DIAMOND")))"""
+    // NOTE: enum trait and enum shape are not 100% compatible. For example, enum trait doesn't support deprecated$since and deprecated$message.
+    val haertWithDeprecationAndDocRendered = List(
```
Comment (i10416) [inline modules/codegen/test/src/smithy4s/codegen/internals/RendererSpec.scala:147]: Here, I use List instead of single String because multiline assertion somehow failed on CI but worked on my local machine.

**Review by Baccata — APPROVED** [review]

**Comment by Baccata** [comment]
Thank you !
